src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/TransportCreation/TransportCreationHandler.cs

```diff
@@ -0,0 +1,192 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient;
+using Microsoft.Data.SqlClient.SNI;
+using Microsoft.Data.SqlClientX.Handlers.Connection;
+
+namespace Microsoft.Data.SqlClientX.Handlers.TransportCreation
+{
+    internal class TransportCreationHandler : IHandler<ConnectionHandlerContext>
+    {
+        private const int KeepAliveIntervalSeconds = 1;
+        private const int KeepAliveTimeSeconds = 30;
+
+        #if NET8_0_OR_GREATER
+        private static readonly TimeSpan DefaultPollTimeout = TimeSpan.FromSeconds(30);
+        #else
+        private const int DefaultPollTimeout = 30 * 100000; // 30 seconds as microseconds
+        #endif
+
+        /// <inheritdoc />
+        public IHandler<ConnectionHandlerContext> NextHandler { get; set; }
+
+        /// <inheritdoc />
+        public async ValueTask Handle(ConnectionHandlerContext context, bool isAsync, CancellationToken ct)
+        {
+            if (context.DataSource is null)
+            {
+                context.Error = new ArgumentNullException(nameof(context));
+                return;
+            }
+
+            try
+            {
+                // @TODO: Build CoR for handling the different protocols in order
+                if (context.DataSource.ResolvedProtocol is DataSource.Protocol.TCP)
+                {
+                    context.ConnectionStream = await HandleTcpRequest(context, isAsync, ct).ConfigureAwait(false);
+                }
+            }
+            catch (Exception e)
+            {
+                context.Error = e;
```

Comment:
Ah, yep, that was actually a bug. Thank you for calling it out.

Regarding the general question here, I raised a similar question and the feedback I got was that it's generally better to return the exception rather than catch it. I'm not 100% sure about this, but it's what we went for so far. You'll likely see this pattern in other PRs for the connection creation code.